File: src/main/java/eu/codetopic/java/utils/log/LogsHandler.java
/*
 * Copyright 2017 Jiří Kuchyňka (Anty)
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package eu.codetopic.java.utils.log;

import java.util.ArrayList;
import java.util.List;

import eu.codetopic.java.utils.ArrayTools;
import eu.codetopic.java.utils.log.base.LogLine;
import eu.codetopic.java.utils.log.base.Priority;

public final class LogsHandler {

    private static final String LOG_TAG = "LogsHandler";

    private final List<OnLoggedListener> listeners = new ArrayList<>();

    LogsHandler() {
    }

    public synchronized void addOnLoggedListener(OnLoggedListener listener) {
        listeners.add(listener);
    }

    public synchronized void removeOnLoggedListener(OnLoggedListener listener) {
        listeners.remove(listener);
    }

    synchronized void onLogged(final LogLine logLine) {
        Priority priority = logLine.getPriority();
        for (OnLoggedListener listener : listeners) {
            Priority[] allowedPriorities = listener.filterPriorities();
            if (allowedPriorities == null || ArrayTools.contains(allowedPriorities, priority)) {
                listener.onLogged(logLine);
            }
        }
    }

    public interface OnLoggedListener {

        Priority[] filterPriorities();

        void onLogged(LogLine logLine);
    }

}

File: src/main/java/eu/codetopic/java/utils/log/LogsHandler.kt
/*
 * javautils
 * Copyright (C)   2017  anty
 *
 * This program is free  software: you can redistribute it and/or modify
 * it under the terms  of the GNU General Public License as published by
 * the Free Software  Foundation, either version 3 of the License, or
 * (at your option) any  later version.
 *
 * This program is distributed in the hope that it  will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied  warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You  should have received a copy of the GNU General Public License
 * along  with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package eu.codetopic.java.utils.log

import eu.codetopic.java.utils.log.base.LogLine
import java.util.*

class LogsHandler internal constructor() {

    companion object {

        private const val LOG_TAG = "LogsHandler"
    }

    private val listeners = WeakHashMap<(LogLine) -> Unit, Unit>()

    @Synchronized
    fun addOnLoggedListener(listener: (LogLine) -> Unit) {
        listeners[listener] = Unit
    }

    @Synchronized
    fun removeOnLoggedListener(listener: (LogLine) -> Unit) {
        listeners.remove(listener)
    }

    @Synchronized
    internal fun onLogged(logLine: LogLine) {
        listeners.keys.forEach { it(logLine) }
    }

}
